libs/common/ring_buffer.hpp

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_COMMON_RING_BUFFER_HPP
+#define IROHA_COMMON_RING_BUFFER_HPP
+
+namespace iroha {
+
+    template<typename T, size_t Count> class RingBuffer final {
+    public:
+        using type = T;
+        using handle = size_t;
+
+    private:
+        static_assert(Count > 0, "Unexpected count value. It must be above 0.");
+
+        enum { kSizeType = sizeof(type) };
+        enum { kActualLimit = Count };
+        enum { kCapLimit = kActualLimit + 1 };
+        enum { kVirtualLimit = 2 * kActualLimit };
+
+        struct Node {
+            uint8_t data[kSizeType];
+        } data_[Count];
+
+        handle end_;
+        handle begin_;
+
+        inline bool handleInBound_(handle register h) const {
+            /**
+             * this code is only for debug purpose
+            **/
+            auto register vbegin = begin_ + kVirtualLimit;
+            h += kVirtualLimit;
+            return ((h >= end_) && (h < vbegin));
+        }
+
+        inline size_t incrementAndNormilize(size_t val) const {
+            return (++val % kVirtualLimit);
+        }
+
+        inline size_t handle2Position(handle h) const {
+            return (h % kActualLimit);
+        }
+
+        inline size_t size_() const {
+            assert(((begin_ - end_) % kVirtualLimit) < kCapLimit);
+            return ((begin_ - end_) % kVirtualLimit);
+        }
+
+        inline bool empty_() const {
+            return (begin_ == end_);
+        }
+
+        inline type& getItem_(Node& node) {
+            return *(type*)node.data;
+        }
+
+        inline type const& getItem_(Node const& node) const {
+            return *(type const*)node.data;
+        }
+
+        inline void destruct_(Node& node) {
```

Comment:
this method works under assertion that `node` refers to the last element, so let's place an assertment
```suggestion
        inline void destructLast_(Node& node) {
          assert(&node == static_cast<Node*>(data_ + handle2Position(end_)));
```

Reply:
No-no-no!!! Why you make a douplicate of the method to get node from handle???

Reply:
The name of the function is fully consistent with its purpose. It must destroy the node that is passed to it. Why should I bind it to the last node? Similarly with construction.

Reply:
the assertion is needed because it is assumed in the following line that the last element is destroyed:
```c++
end_ = incrementAndNormilize(end_);
```

Reply:
I agree, assert is definitely needed. It must be placed in a method that determines the position of the deleted node.

Reply:
You are right, there should not be a determining position in this method ...